extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -368,40 +371,87 @@ export abstract class AzureAuth implements vscode.Disposable {
 				allSubs.push(...subscriptions);
 			} catch (ex) {
 				console.log(ex);
-				throw new Error('Error retreiving subscription information');
+				throw new Error('Error retrieving subscription information');
 			}
 		}
 		return allSubs;
 	}
 
-	protected async getToken(postData: { [key: string]: string }, tenant = this.commonTenant, resourceId: string = ''): Promise<TokenRefreshResponse | undefined> {
+	protected async getToken(postData: { [key: string]: string }, tenant = this.commonTenant, resourceId: string = '', resourceEndpoint: string = ''): Promise<TokenRefreshResponse | undefined> {
 		try {
-			const tokenUrl = `${this.loginEndpointUrl}${tenant}/oauth2/token`;
-
-			const tokenResponse = await this.makePostRequest(tokenUrl, postData);
-			const tokenClaims = this.getTokenClaims(tokenResponse.data.access_token);
-
-			const accessToken: AccessToken = {
-				token: tokenResponse.data.access_token,
-				key: tokenClaims.email || tokenClaims.unique_name || tokenClaims.name,
-			};
-
-			this.memdb.set(this.createMemdbString(accessToken.key, tenant, resourceId), tokenResponse.data.expires_on);
+			let refreshResponse: TokenRefreshResponse;
 
-			const refreshToken: RefreshToken = {
-				token: tokenResponse.data.refresh_token,
-				key: accessToken.key
-			};
-
-			return { accessToken, refreshToken, tokenClaims };
+			try {
+				const tokenUrl = `${this.loginEndpointUrl}${tenant}/oauth2/token`;
+				const tokenResponse = await this.makePostRequest(tokenUrl, postData);
+				const tokenClaims = this.getTokenClaims(tokenResponse.data.access_token);
+
+				const accessToken: AccessToken = {
+					token: tokenResponse.data.access_token,
+					key: tokenClaims.email || tokenClaims.unique_name || tokenClaims.name,
+				};
+
+				const refreshToken: RefreshToken = {
+					token: tokenResponse.data.refresh_token,
+					key: accessToken.key
+				};
+				const expiresOn = tokenResponse.data.expires_on;
+
+				refreshResponse = { accessToken, refreshToken, tokenClaims, expiresOn };
+			} catch (ex) {
+				if (ex?.response?.data?.error === 'interaction_required') {
+					const shouldOpenLink = await this.openConsentDialog(tenant, resourceId);
+					if (shouldOpenLink === true) {
+						const { tokenRefreshResponse, authCompleteDeferred } = await this.promptForConsent(resourceEndpoint, tenant);
+						refreshResponse = tokenRefreshResponse;
+						authCompleteDeferred.resolve();
+					} else {
+						vscode.window.showInformationMessage(localize('azure.noConsentToReauth', "The authentication failed since ADS was unable to open re-authentication page."));
```

Comment:
This message doesn't seem to indicate what happened here to me - isn't the real issue is that they didn't accept the consent request?

Reply:
Essentially I don't want it to be a jarring experience of you're working with ADS and suddenly your browser opens to a sign in page .

I want to ask the user permission before opening it saying hey we're about to open a browser to sign you in, are you okay with that.

Reply:
Yeah so that's fine and I agree - I just don't know if the information message here really meshes with what happened. Because in this case the user specifically said "no, I don't want to open this page" right? In that case do we even need to pop up a separate error? Couldn't the consent dialog just say "if you don't do this you won't be able to authenticate"?

Reply:
Yeah I have no opinions on the wording of this. I'll try to get PM looped in to tell me too :)